system_setup/common/conf.py

```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env python3
+# Copyright 2015-2021 Canonical, Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+#    original code from ubuntuwslctl.core.loader
+#    Copyright (C) 2021 Canonical Ltd.
+
+import logging
+from configparser import ConfigParser
+
+log = logging.getLogger("system_setup.common.conf")
+
+
+class WSLConfig:
+    def __init__(self, inst_type,  conf_file):
+        self.inst_type = inst_type
+        self.conf_file = conf_file
+
+        self.config = ConfigParser()
+        self.config.BasicInterpolcation = None
+        self.config.read(conf_file)
+
+    def update(self, config_section, config_setting, config_value):
+        self.config[config_section][config_setting] = config_value
+        with open(self.conf_file, 'w') as configfile:
+            self.config.write(configfile)
+
+
+class WSLConfigHandler:
```

Comment:
This is neat, but I think you are navigating between generic and non generic code here.

For instance, you want to be generic with a general handler which handles both "wsl" and "ubuntu" configuration. However, the code below
```
        if type_input == "ubuntu":
            return self.ubuntu_conf
        elif type_input == "wsl":
            return self.wsl_conf
        else:
            raise ValueError("Invalid config name. Please check again.")
```
Show that you are indeed restricted to a single set of config.

To avoid too much code (like _select_config, _update) and subsequent file updates (you are updating here each file multiple times in a row), I suggest turning WSLConfigHandler in a serializer and deserializer for wslconf_base and wslconf_ad. It doesn’t need to have any instance.
This means:
- This class has an update() method that takes a `wslconf_base` and `wslconf_ad` parameters directly.
- Then, it splits this up in section and updates both file, each one a single time. (this removes the need for `_select_config` and `_update` as this is the class which knows which property goes to which file. It will simplify also udpate.
dest = "" with a switch statement where all method properties are listed, which then assign to a particular dest.
- Then, this will update in one shot, one for each files, anything loaded on the model.

This should simplify this class in term of metacode, and ensure that only one place will know which keys goes where.

Then, in a separate PR or in the same if you want, when initialize both model, you use this class to load() and return both models.

There are things to think about if a key is the default model value, remove it from the file (don’t write unecessary key. This is what happens with this code)

Reply:
gotcha, I will do a mockup today

Reply:
Can you checkout my new implementation for this? Thanks

Reply:
(see new comments below)